File: 4.controls/mobile/Engine/Source/Objects/CCRenderable.cpp
/*-----------------------------------------------------------
 * 2c - Cross Platform 3D Application Framework
 *-----------------------------------------------------------
 * Copyright © 2010 – 2011 France Telecom
 * This software is distributed under the Apache 2.0 license,
 * see the "license.txt" file for more details.
 *-----------------------------------------------------------
 * File Name   : CCRenerable.h
 *-----------------------------------------------------------
 */

#include "CCDefines.h"
#include "CCRenderable.h"


CCRenderable::CCRenderable()
{
    shouldRender = true;
    dirtyModelMatrix();
	
    updateWorldMatrix = true;
	
	positionPtr = &position;
	scale = NULL;
}


void CCRenderable::destruct()
{
}


void CCRenderable::dirtyModelMatrix()
{
    updateModelMatrix = true;
}


void CCRenderable::dirtyWorldMatrix()
{
	updateWorldMatrix = true;
}


void CCRenderable::refreshModelMatrix()
{
	if( updateModelMatrix )
	{
		CCMatrixLoadIdentity( modelMatrix );	
        CCMatrixTranslate( modelMatrix, position.x, position.y, position.z );
			
		if( scale != NULL )
		{	
			CCMatrixScale( modelMatrix, scale->x, scale->y, scale->z );
		}
		
		if( rotation.x != 0.0f )
		{
            CCMatrixRotate( modelMatrix, rotation.x, 1.0f, 0.0f, 0.0f );
		}
		
		if( rotation.y != 0.0f )
		{
            CCMatrixRotate( modelMatrix, rotation.y, 0.0f, 1.0f, 0.0f );
		}
		
		if( rotation.z != 0.0f )
		{
            CCMatrixRotate( modelMatrix, rotation.z, 0.0f, 0.0f, 1.0f );
		}
			
		updateModelMatrix = false;
		dirtyWorldMatrix();
	}
}


void CCRenderable::setPosition(const CCVector3 &vector)
{
    setPositionXYZ( vector.x, vector.y, vector.z );
}


void CCRenderable::setPositionXYZ(const float x, const float y, const float z)
{
	position.x = x;
	position.y = y;
	position.z = z;
    dirtyModelMatrix();
}


void CCRenderable::setPositionX(const float x)
{
	setPositionXYZ( x, position.y, position.z );
}


void CCRenderable::setPositionXZ(const float x, const float z)
{
	setPositionXYZ( x, position.y, z );
}


void CCRenderable::translate(CCVector3 *vector)
{
	translate( vector->x, vector->y, vector->z );
}


void CCRenderable::translate(const float x, const float y, const float z)
{
	position.x += x;
	position.y += y;
	position.z += z;
    dirtyModelMatrix();
}


void CCRenderable::rotationUpdated()
{
    dirtyModelMatrix();
}


void CCRenderable::setRotation(const CCVector3 &vector)
{
    rotation = vector;
    rotationUpdated();
}


void CCRenderable::setRotationY(const float y)
{
    rotation.y = y;
    rotationUpdated();
}


void CCRenderable::rotateX(const float x)
{
	rotation.x += x;
    CCClampRotation( rotation.x );
    rotationUpdated();
}


void CCRenderable::rotateY(const float y)
{
	rotation.y += y;
    CCClampRotation( rotation.y );
    rotationUpdated();
}


void CCRenderable::rotateZ(const float z)
{
	rotation.z += z;
    CCClampRotation( rotation.z );
    rotationUpdated();
}

File: 3.cameraandmatrices/webgl/source/js/engine/objects/CCRenderable.js
/*-----------------------------------------------------------
 * 2c - Cross Platform 3D Application Framework
 *-----------------------------------------------------------
 * Copyright © 2011 - 2011 France Telecom
 * This software is distributed under the Apache 2.0 license.
 * http://www.apache.org/licenses/LICENSE-2.0.html
 *-----------------------------------------------------------
 * File Name   : CCRenderable.js
 * Description : A renderable component.
 *
 * Created     : 23/10/11
 * Author(s)   : Ashraf Samy Hegab
 *-----------------------------------------------------------
 */

function CCRenderable()
{
	this.construct();
}


CCRenderable.prototype.construct = function()
{
    this.modelMatrix = mat4.create();
    this.position = vec3.create();
    this.rotation = vec3.create();
    this.scale = vec3.create( [1,1,1] );
	
	this.shouldRender = true;
    this.dirtyModelMatrix();
}


CCRenderable.prototype.destruct = function()
{
}


CCRenderable.prototype.dirtyModelMatrix = function()
{
	this.updateModelMatrix = true;
}


CCRenderable.prototype.refreshModelMatrix = function()
{
	if( this.updateModelMatrix )
    {        
        var modelMatrix = this.modelMatrix;
        mat4.identity( modelMatrix );
        mat4.translate( modelMatrix, this.position );

        mat4.scale( modelMatrix, this.scale );

        var rotation = this.rotation;
        mat4.rotate( modelMatrix, rotation[0], [1.0, 0.0, 0.0] );
        mat4.rotate( modelMatrix, rotation[1], [0.0, 1.0, 0.0] );
        mat4.rotate( modelMatrix, rotation[2], [0.0, 0.0, 1.0] );
		
        this.updateModelMatrix = false;
    }
}


CCRenderable.prototype.setPosition = function(vector)
{
    this.setPositionXYZ( vector[0], vector[1], vector[2] );
}


CCRenderable.prototype.setPositionXYZ = function(x, y, z)
{
    var position = this.position;
    position[0] = x;
    position[1] = y;
    position[2] = z;
    this.updateModelMatrix = true;
    this.updateCollisions = true;
    this.refreshCollisions();
}


CCRenderable.prototype.setPositionX = function(x)
{
	this.setPositionXYZ( x, this.position[1], this.position[2] );
}


CCRenderable.prototype.translate = function(x, y, z)
{
	var position = this.position;
    position[0] += x;
	position[1] += y;
	position[2] += z;
    this.updateModelMatrix = true;
    this.updateCollisions = true;
    this.refreshCollisions();
}


CCRenderable.prototype.refreshCollisions = function()
{
}
